README.md
# YunoHost App for Hubzilla Hub

## Hubzilla
[Hubzilla](http://hubzilla.org) is a powerful platform for creating interconnected websites featuring a decentralized identity, communications, and permissions framework built using common webserver technology. 


Current snapshot in *sources*: 

* https://github.com/redmatrix/hubzilla: 2.0.7
* https://github.com/redmatrix/hubzilla-addons: >2.0 (commit 33a9a7f971097bcf14228b626bfb127559e47830)

## Important Notes

Before installing, read the [Hubzilla installation instructions](https://github.com/redmatrix/hubzilla/blob/master/install/INSTALL.txt) for important information about 

- SSL certificate validation requirement (now with support for [Let's Encrypt!](https://letsencrypt.org)). See Installation section below.
- Dedicated domain (must install under web root like **https://hub.example.com/** not **https://example.com/hub/** )
- Required packages (all of these are not yet installed by this YunoHost installer package). This YunoHost package installs the following additional packages:
  - php5-cli 
  - php5-imagick 
  - php5-gd 
  - php5-mcrypt 
- This package requires a **system-wide change to php.ini** that enables the `exec()` perimission. [See the PHP manual for more information](php.net/manual/function.exec.php).



## Installation

### Register a new domain and add it to YunoHost
Hubzilla requires a dedicated domain, so obtain one and add it using the [YunoHost admin](https://reticu.li/yunohost/admin) panel. **Domains -> Add domain**

Once you have added the new domain to YunoHost, SSH into your YunoHost server and perform the following steps:

1. Install [certbot](https://certbot.eff.org/) to make installing free SSL certificates from Let's Encrypt simple.

1. Stop nginx

		service nginx stop
		
1. Run the **certbot** utility with the **certonly** option

		certbot certonly

1. Copy the generated certificate and key into the appropriate location for YunoHost to use

		cp /etc/letsencrypt/live/YOUR_DOMAIN/fullchain.pem /etc/yunohost/certs/YOUR_DOMAIN/crt.pem 
		cp /etc/letsencrypt/live/YOUR_DOMAIN/privkey.pem /etc/yunohost/certs/YOUR_DOMAIN/key.pem 

1. Restart nginx
		
		service nginx start
		
### Install the Hubzilla application
Use the [YunoHost admin](https://reticu.li/yunohost/admin) panel to install Hubzilla by entering the GitHub repo address in the custom app URL

		https://github.com/YunoHost-Apps/hubzilla-yunohost

Make sure to select your domain from the previous section as the application domain. Also set the application to Public.

When installation is complete, you will need to visit your new hub and register a new account using the email address you specified in the app installation form. You should then be able to log in and create your first channel.

sources/CHANGELOG
Hubzilla 2.0 (2016-12-23)
	- Deprecate bb_iframe
	- Note widget: resize the textarea to reveal full content
	- Implement fixed left aside
	- Implement lockview for wikilist
	- Simplify wikilist widget
	- Router error reporting
	- Setup changes to check for shell_exec and exec functions
	- Extensible permissions upgrade handling for channels with custom permission roles
	- Allow plugins to cancel item_store() and item_store_update()
	- ZOT version 1.2 provides negotiation of cryptographic algorithms
	- Provide a fresh new look and cleaner layout and more relevant information to siteinfo
	- Introduce highlight bbcode [hl]
	- Implement wiki mimetypes markdown or bbcode
	- Doc pages refactoring
	- Update webpages and wiki context help
	- Make a git commit when a new wiki page is created
	- Prev-next navigation for mod_connedit to ease bulk connection edits
	- Move the remote user homebutton to the user menu
	- Do not render maps/locations for Diaspora destinations
	- Provide 'per-page' caching for is_matrix_url() results to reduce duplicate queries
	- Don't send notification for posts/comments on old conversations that were refetched after having expired
	- Numerous wiki UI improvements
	- Move twitter api to addon
	- Cleanup and re-organise the voting and attendance buttons
	- Reorganise emoticons
	- Collapse navbar-collapse-1 if avatar menu is clicked.
	- New display setting: static page update as opposed to live update
	- Command line administrative channel connect utility
	- Modernise chanview
	- Implement edit activities to share post/comment edits with protocols which do not support them (e.g. Diaspora)
	- Wiki export
	- Numerous postgres compatibility fixes
	- Remove requirement that imported profile photos be in the profile photos album
	- Change event behaviour - share by default.
	- Use PDO database driver exclusively (deprecate drivers that are separately maintained)
	- Zot API re-write and extended

	Bugfixes
	- Fix z_fetch_url() incorrect variable
	- Fix SQL error with app categories
	- Fix do not show revert buttons if we do not have write perms
	- Fix dropdown positions
	- Fix do not increase opacity to more than 1
	- Fix clone sync missing for some item delete operations
	- Fix embed-image for fullscreen mode
	- Fix attach_list_files()
	- Fix full screen for embedded videos
	- Fix the forum widget for forums with custom perms
	- Fix issue #607 parens not recognised inside urls
	- Fix pubsites: don't list dead sites
	- Fix issue #596 silence headers already sent warning
	- Fix missing plugins in zot-info
	- Fix notification issue
	- Fix issue #594 like of thing appears as profile owner like
	- Fix export issue
	- Fix checklist bbcode - only turn [] and [x] into checkboxes if it is found inside a checklist
	- Fix wiki permissions issues
	- Fix public calendar leaks connection information (birthdays) when view_contacts is not allowed
	- Fix attach_rename: flaw in duplicate filename detection resulted in filename(1)(1)(1).ext
	- Fix a fatal error with incorrect DB object access
	- Provide /locs link on settings page if there is more than one hubloc for this channnel *that isn't deleted*.
	- Fix issue #577 if connecting to a channel that is already pending, undo the pending and set connect permissions accordingly
	- Fix issue #575, when 'nofinish' is set on an event, invalid date was generated/stored
	- Fix bbcode event formatting issue
	- Fix zot_finger from navbar people search looping
	- Fix fromStandalonePermission()

	Plugins
	- GNU Social: removed from addons for security reasons - it might be re-implemented once it is properly reviewed
	- Diaspora: missing item author when diaspora public comment received from relay
	- Superblock: refactoring
	- New addon: tripleaes for pro
	- Cdav:  "if not exists" only supported starting with postgresql v. 9.5 debian stable has 9.4
	- Rendezvous: added markers and members export tool at /rendezvous/[group_id]/export/{markers,members}
	- Twitter: move twitter api to addon
	- New addon: b2tbtn (back to top button)
	- Diaspora: import public diaspora messages to sys if applicable
	- Diaspora: try and handle singletons better and simplify the associated notifier decisions
	- Rendezvous: add proximity alert feature to members to issue notification when member is within a specified distance.
	- New addon: diaspora_reconnect to refriend diaspora/friendica connections from a clone or channel move
	- Diaspora: change the logic for deciding between upstream and downstream message flow for notifier plugins
	- Rendezvous: prompt member to share their location by activating the GPS control using a tooltip and pulsing visibility
	- statistics_json: fix nodeinfo
	- Rendezvous: restored the lost gps-icon.png and corrected the OpenStreetMap tile server URL to avoid insecure content warnings
	- Rendezvous: use observer name if available
	- std_embeds: missing backslash
	- Diaspora: postgres fixes issue #31
	- Rendezvous: added marker list with centering buttons and popup open.
	- Rendezvous: added control to see list of members sharing their location, with buttons to pan the map to center them
	- Diaspora: system level diaspora toggle
	- Rendezvous: added control that displays members.
	- Diaspora: rename diaspora2bb() to markdown_to_bb() in core
	- Hubwall: remove illegal unescaped angle chars
	- Rendezvous: Add control to delete member if not updated in over 14 minutes

Hubzilla 1.14 (2016-10-13)
	- New hook bbcode_filter
	- Unify the various mail sending instance to enotify::send() and z_mail()
	- Provide ability for admin to change account password
	- Replace deprecated Sabre functions
	- Add plugin hook for 'get_profile_photo'
	- Convert NULL_DATE to a legal date for compatibility with MySQL strict mode
	- Allow a site to over-ride the help table-of-contents files
	- Autoscroll to target post/comment when in single-thread mode
	- Indicator for own response verb activity
	- Add server role documentation

sources/mod/cal.php
<?php

require_once('include/conversation.php');
require_once('include/bbcode.php');
require_once('include/datetime.php');
require_once('include/event.php');
require_once('include/items.php');
require_once('include/Contact.php');


function cal_init(&$a) {
	if((get_config('system','block_public')) && (! local_channel()) && (! remote_channel())) {
		return;
	}

	$o = '';

	if(argc() > 1) {
		$nick = argv(1);

		profile_load($a,$nick);

		$channelx = channelx_by_nick($nick);

		if(! $channelx)
			return;

		App::$data['channel'] = $channelx;

		$observer = App::get_observer();
		App::$data['observer'] = $observer;

		$observer_xchan = (($observer) ? $observer['xchan_hash'] : '');

		head_set_icon(App::$data['channel']['xchan_photo_s']);

		App::$page['htmlhead'] .= "<script> var ispublic = '" . t('everybody') . "'; var profile_uid = " . ((App::$data['channel']) ? App::$data['channel']['channel_id'] : 0) . "; </script>" ;

	}

	return;
}



function cal_content(&$a) {

	if((get_config('system','block_public')) && (! local_channel()) && (! remote_channel())) {
		return;
	}


	$channel = null;

	if(argc() > 1) {
		$channel = channelx_by_nick(argv(1));
	}


	if(! $channel) {
		notice( t('Channel not found.') . EOL);
		return;
	}

	// since we don't currently have an event permission - use the stream permission

	if(! perm_is_allowed($channel['channel_id'], get_observer_hash(), 'view_stream')) {
		notice( t('Permissions denied.') . EOL);
		return;
	}

	$sql_extra = permissions_sql($channel['channel_id'],get_observer_hash(),'event');

	$first_day = get_pconfig(local_channel(),'system','cal_first_day');
	$first_day = (($first_day) ? $first_day : 0);

	$htpl = get_markup_template('event_head.tpl');
	App::$page['htmlhead'] .= replace_macros($htpl,array(
		'$baseurl' => z_root(),
		'$module_url' => '/cal/' . $channel['channel_address'],
		'$modparams' => 2,
		'$lang' => App::$language,
		'$first_day' => $first_day
	));

	$o = '';

	$tabs = profile_tabs($a, True, $channel['channel_address']);

	$mode = 'view';
	$y = 0;
	$m = 0;
	$ignored = ((x($_REQUEST,'ignored')) ? " and ignored = " . intval($_REQUEST['ignored']) . " "  : '');

	// logger('args: ' . print_r(App::$argv,true));

	if(argc() > 3 && intval(argv(2)) && intval(argv(3))) {
		$mode = 'view';
		$y = intval(argv(2));
		$m = intval(argv(3));
	}
	if(argc() <= 3) {
		$mode = 'view';
		$event_id = argv(2);
	}

	if($mode == 'view') {

		/* edit/create form */
		if($event_id) {
			$r = q("SELECT * FROM `event` WHERE event_hash = '%s' AND `uid` = %d LIMIT 1",
				dbesc($event_id),
				intval($channel['channel_id'])
			);
			if(count($r))
				$orig_event = $r[0];
		}


		// Passed parameters overrides anything found in the DB
		if(!x($orig_event))
			$orig_event = array();



		$tz = date_default_timezone_get();
		if(x($orig_event))
			$tz = (($orig_event['adjust']) ? date_default_timezone_get() : 'UTC');

		$syear = datetime_convert('UTC', $tz, $sdt, 'Y');
		$smonth = datetime_convert('UTC', $tz, $sdt, 'm');
		$sday = datetime_convert('UTC', $tz, $sdt, 'd');
		$shour = datetime_convert('UTC', $tz, $sdt, 'H');
		$sminute = datetime_convert('UTC', $tz, $sdt, 'i');

		$stext = datetime_convert('UTC',$tz,$sdt);
		$stext = substr($stext,0,14) . "00:00";

		$fyear = datetime_convert('UTC', $tz, $fdt, 'Y');
		$fmonth = datetime_convert('UTC', $tz, $fdt, 'm');
		$fday = datetime_convert('UTC', $tz, $fdt, 'd');
		$fhour = datetime_convert('UTC', $tz, $fdt, 'H');
		$fminute = datetime_convert('UTC', $tz, $fdt, 'i');

		$ftext = datetime_convert('UTC',$tz,$fdt);
		$ftext = substr($ftext,0,14) . "00:00";

		$type = ((x($orig_event)) ? $orig_event['type'] : 'event');

		$f = get_config('system','event_input_format');
		if(! $f)
			$f = 'ymd';

		$catsenabled = feature_enabled($channel['channel_id'],'categories');


		$show_bd = perm_is_allowed($channel['channel_id'], get_observer_hash(), 'view_contacts');
		if(! $show_bd) {
			$sql_extra .= " and event.type != 'birthday' ";
		}


		$category = '';

		$thisyear = datetime_convert('UTC',date_default_timezone_get(),'now','Y');
		$thismonth = datetime_convert('UTC',date_default_timezone_get(),'now','m');
		if(! $y)
			$y = intval($thisyear);
		if(! $m)
			$m = intval($thismonth);

		// Put some limits on dates. The PHP date functions don't seem to do so well before 1900.
		// An upper limit was chosen to keep search engines from exploring links millions of years in the future. 

		if($y < 1901)
			$y = 1900;
		if($y > 2099)
			$y = 2100;

		$nextyear = $y;
		$nextmonth = $m + 1;
		if($nextmonth > 12) {
				$nextmonth = 1;
			$nextyear ++;
		}

		$prevyear = $y;
		if($m > 1)
			$prevmonth = $m - 1;
		else {
			$prevmonth = 12;
			$prevyear --;
		}
			
		$dim    = get_dim($y,$m);
		$start  = sprintf('%d-%d-%d %d:%d:%d',$y,$m,1,0,0,0);
		$finish = sprintf('%d-%d-%d %d:%d:%d',$y,$m,$dim,23,59,59);


		if (argv(2) === 'json'){
			if (x($_GET,'start'))	$start = $_GET['start'];
			if (x($_GET,'end'))	$finish = $_GET['end'];
		}

		$start  = datetime_convert('UTC','UTC',$start);
		$finish = datetime_convert('UTC','UTC',$finish);

		$adjust_start = datetime_convert('UTC', date_default_timezone_get(), $start);
		$adjust_finish = datetime_convert('UTC', date_default_timezone_get(), $finish);

		if (x($_GET,'id')){
		  	$r = q("SELECT event.*, item.plink, item.item_flags, item.author_xchan, item.owner_xchan
                                from event left join item on resource_id = event_hash where resource_type = 'event' and event.uid = %d and event.id = %d $sql_extra limit 1",
				intval($channel['channel_id']),
				intval($_GET['id'])
			);
		} 
		else {
			// fixed an issue with "nofinish" events not showing up in the calendar.
			// There's still an issue if the finish date crosses the end of month.
			// Noting this for now - it will need to be fixed here and in Friendica.
			// Ultimately the finish date shouldn't be involved in the query. 

			$r = q("SELECT event.*, item.plink, item.item_flags, item.author_xchan, item.owner_xchan
                              from event left join item on event_hash = resource_id 
				where resource_type = 'event' and event.uid = %d $ignored 
				AND (( adjust = 0 AND ( finish >= '%s' or nofinish = 1 ) AND start <= '%s' ) 
				OR  (  adjust = 1 AND ( finish >= '%s' or nofinish = 1 ) AND start <= '%s' )) $sql_extra ",
				intval($channel['channel_id']),
				dbesc($start),
				dbesc($finish),
				dbesc($adjust_start),
				dbesc($adjust_finish)
			);

		}

		$links = array();

		if($r) {
			xchan_query($r);
			$r = fetch_post_tags($r,true);

			$r = sort_by_date($r);
		}

		if($r) {
			foreach($r as $rr) {
				$j = (($rr['adjust']) ? datetime_convert('UTC',date_default_timezone_get(),$rr['start'], 'j') : datetime_convert('UTC','UTC',$rr['start'],'j'));
				if(! x($links,$j)) 
					$links[$j] = z_root() . '/' . App::$cmd . '#link-' . $j;
			}
		}

		$events=array();

		$last_date = '';
		$fmt = t('l, F j');

		if($r) {

			foreach($r as $rr) {
				
				$j = (($rr['adjust']) ? datetime_convert('UTC',date_default_timezone_get(),$rr['start'], 'j') : datetime_convert('UTC','UTC',$rr['start'],'j'));
				$d = (($rr['adjust']) ? datetime_convert('UTC',date_default_timezone_get(),$rr['start'], $fmt) : datetime_convert('UTC','UTC',$rr['start'],$fmt));
				$d = day_translate($d);
				
				$start = (($rr['adjust']) ? datetime_convert('UTC',date_default_timezone_get(),$rr['start'], 'c') : datetime_convert('UTC','UTC',$rr['start'],'c'));
				if ($rr['nofinish']){
					$end = null;
				} else {
					$end = (($rr['adjust']) ? datetime_convert('UTC',date_default_timezone_get(),$rr['finish'], 'c') : datetime_convert('UTC','UTC',$rr['finish'],'c'));
				}
				
				
				$is_first = ($d !== $last_date);
					
				$last_date = $d;

				$edit = false;

				$drop = false;

				$title = strip_tags(html_entity_decode(bbcode($rr['summary']),ENT_QUOTES,'UTF-8'));
				if(! $title) {
					list($title, $_trash) = explode("<br",bbcode($rr['desc']),2);
					$title = strip_tags(html_entity_decode($title,ENT_QUOTES,'UTF-8'));
				}
				$html = format_event_html($rr);
				$rr['desc'] = bbcode($rr['desc']);
				$rr['location'] = bbcode($rr['location']);
				$events[] = array(
					'id'=>$rr['id'],
					'hash' => $rr['event_hash'],
					'start'=> $start,
					'end' => $end,
					'drop' => $drop,
					'allDay' => false,
					'title' => $title,
					
					'j' => $j,
					'd' => $d,
					'edit' => $edit,
					'is_first'=>$is_first,
					'item'=>$rr,
					'html'=>$html,
					'plink' => array($rr['plink'],t('Link to Source'),'',''),
				);


			}
		}
		
		if (argv(2) === 'json'){
			echo json_encode($events); killme();
		}
		
		// links: array('href', 'text', 'extra css classes', 'title')
		if (x($_GET,'id')){
			$tpl =  get_markup_template("event_cal.tpl");
		} 
		else {
			$tpl = get_markup_template("events_cal-js.tpl");
		}

		$nick = $channel['channel_address'];

		$o = replace_macros($tpl, array(
			'$baseurl'	=> z_root(),
			'$new_event'	=> array(z_root().'/cal',(($event_id) ? t('Edit Event') : t('Create Event')),'',''),
			'$previus'	=> array(z_root()."/cal/$nick/$prevyear/$prevmonth",t('Previous'),'',''),
			'$next'		=> array(z_root()."/cal/$nick/$nextyear/$nextmonth",t('Next'),'',''),
			'$export'	=> array(z_root()."/cal/$nick/$y/$m/export",t('Export'),'',''),
			'$calendar'	=> cal($y,$m,$links, ' eventcal'),
			'$events'	=> $events,
			'$upload'	=> t('Import'),
			'$submit'	=> t('Submit'),
			'$prev'		=> t('Previous'),
			'$next'		=> t('Next'),
			'$today'	=> t('Today'),
			'$form'		=> $form,
			'$expandform'	=> ((x($_GET,'expandform')) ? true : false),
			'$tabs'		=> $tabs
		));
		
		if (x($_GET,'id')){ echo $o; killme(); }
		
		return $o;
	}

}
